ttnn/cpp/pybind11/pytensor.cpp

```diff
@@ -260,187 +182,182 @@ Tensor convert_torch_tensor_to_tt_tensor(
-    // Otherwise, figure it out from numpy dtype
     DataType data_type;
-    if (optional_data_type.has_value()) {
-        data_type = optional_data_type.value();
-    } else if (np_dtype.equal(np.attr("float32"))) {
-        data_type = DataType::FLOAT32;
-    } else if (np_dtype.equal(np.attr("float16"))) {
-        contiguous_np_tensor = contiguous_np_tensor.attr("astype")(np.attr("float32"));
-        // TODO(arakhmati): add DataType::FLOAT16?
-        data_type = DataType::BFLOAT16;
-    } else if (np_dtype.equal(np.attr("int64"))) {
-        contiguous_np_tensor = contiguous_np_tensor.attr("astype")(np.attr("int32"));
-        // TODO(arakhmati): add DataType::INT64?
-        data_type = DataType::UINT32;
-    } else if (np_dtype.equal(np.attr("int32"))) {
-        // TODO(arakhmati): add DataType::INT32?
-        data_type = DataType::UINT32;
-    } else if (np_dtype.equal(np.attr("ubyte"))) {
-        data_type = DataType::UINT8;
-    } else {
-        TT_THROW("Unsupported DataType: {}", std::string(py::repr(np_dtype)));
-    }
 
-    switch (data_type) {
-        case DataType::UINT8: {
-            if (not np_dtype.equal(np.attr("ubyte"))) {
-                contiguous_np_tensor = contiguous_np_tensor.attr("astype")(np.attr("ubyte"));
-            }
-            break;
+    py::object contiguous_py_tensor;
+    size_t num_elements = 0;
+    size_t py_data_ptr = 0;
+    if (py::isinstance(py_tensor, torch.attr("Tensor"))) {
+        contiguous_py_tensor = py_tensor.attr("contiguous")();
+
+        // Override the data type if there is an user-provided one
+        // Otherwise, figure it out from torch dtype
+        if (optional_data_type.has_value()) {
+            data_type = optional_data_type.value();
+        } else if (py_dtype.equal(torch.attr("float32"))) {
+            data_type = DataType::FLOAT32;
+        } else if (py_dtype.equal(torch.attr("float16"))) {
+            data_type = DataType::BFLOAT16;
+        } else if (py_dtype.equal(torch.attr("bfloat16"))) {
+            data_type = DataType::BFLOAT16;
+        } else if (py_dtype.equal(torch.attr("int64"))) {
+            // TODO: add DataType::INT64?
+            data_type = DataType::UINT32;
+        } else if (py_dtype.equal(torch.attr("int32"))) {
+            data_type = DataType::INT32;
+        } else if (py_dtype.equal(torch.attr("int16"))) {
+            // TODO: add DataType::INT16?
+            data_type = DataType::UINT16;
+        } else if (py_dtype.equal(torch.attr("uint8"))) {
+            data_type = DataType::UINT8;
+        } else {
+            TT_THROW("Unsupported DataType: {}", std::string(py::repr(py_dtype)));
         }
-        case DataType::UINT16: {
-            if (not np_dtype.equal(np.attr("int32"))) {
-                contiguous_np_tensor = contiguous_np_tensor.attr("astype")(np.attr("int16"));
+
+        switch (data_type) {
+            case DataType::UINT8: {
+                if (not py_dtype.equal(torch.attr("uint8"))) {
```

Comment:
Let's create a lambda out of this snippet, wdyt?

```
auto maybe_convert_tensor = [&contiguous_py_tensor](std::string_view pytorch_type) {
  if (not py_dtype.equal(torch.attr(pytorch_type))) {
    contiguous_py_tensor = contiguous_py_tensor.attr("to")(torch.attr(pytorch_type));
  }
};
```

Reply:
good suggestion!